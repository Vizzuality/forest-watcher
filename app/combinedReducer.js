import * as reducers from 'redux-modules';
import { combineReducers } from 'redux';

/**
 * Separated so we can test the store state without having to mock loads of dependencies.
 */
export const combinedReducer = combineReducers(reducers);
