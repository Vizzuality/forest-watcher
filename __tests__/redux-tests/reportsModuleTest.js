import 'react-native';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { combinedReducer } from 'combinedReducer';
import reportsReducer, {
  createReport,
  deleteReport,
  getDefaultReport,
  saveReport,
  setReportAnswer,
  syncReports,
  uploadReport
} from 'redux-modules/reports';

describe('Redux Reports Module', () => {
  const mockArea = {
    name: 'nameMock',
    id: 'areaIDMock',
    application: 'applicationMock', // used to test that all fields are included in payload
    geostore: { id: 'geostoreIDMock' }
  };

  const mockCreateReport = {
    reportName: 'mockCreateReportName',
    userPosition: '1,2',
    clickedPosition: '3,4',
    area: mockArea
  };

  const mockReport = {
    name: 'reportNameMock',
    area: mockArea,
    userPosition: '1,2',
    clickedPosition: '3,4',
    index: 1,
    status: 'draft',
    date: 'mockDateString',
    answers: []
  };

  const mockAnswerChild = {
    questionName: 'mockQuestionName2',
    value: 'Value2'
  };

  const mockAnswerParent = {
    questionName: 'mockQuestionName1',
    value: 'Value1',
    child: mockAnswerChild
  };

  const mockReportWithAnswers = {
    ...mockReport,
    answers: [mockAnswerParent]
  };

  it('Initial reducer state', () => {
    expect(reportsReducer(undefined, { type: 'NONE' })).toMatchSnapshot();
  });

  describe('Simple Actions: snapshot and reducer test', () => {
    // Snapshot tests the action object and the reducer state
    function simpleActionTest(action, propertyMatcher) {
      const actionPropertyMatcher = propertyMatcher ? { payload: propertyMatcher } : {};
      const statePropertyMatcher = propertyMatcher ? { list: propertyMatcher } : {};
      expect(action).toMatchSnapshot(actionPropertyMatcher);
      expect(reportsReducer(undefined, action)).toMatchSnapshot(statePropertyMatcher);
    }

    it('createReport', () => {
      const propertyMatcher = { mockCreateReportName: { date: expect.any(String) } };
      simpleActionTest(createReport(mockCreateReport), propertyMatcher);
    });

    it('deleteReport', () => {
      simpleActionTest(deleteReport('mockReportName'));
    });

    it('getDefaultReport', () => {
      simpleActionTest(getDefaultReport());
    });

    it('saveReport', () => {
      simpleActionTest(saveReport(mockReport.name, mockReport));
    });

    it('saveReport with answers', () => {
      simpleActionTest(saveReport(mockReportWithAnswers.name, mockReportWithAnswers));
    });

    it('setReportAnswer', () => {
      simpleActionTest(setReportAnswer('mockReportName', mockAnswerParent, true));
      // simpleActionTest(setReportAnswer('mockReportName', mockAnswerParent, false));
    });
  });

  describe('Redux Snapshot Thunk Actions', () => {
    let initialStoreState;
    let configuredStore;
    let store;

    beforeAll(() => {
      // create store
      initialStoreState = combinedReducer(undefined, { type: 'NONE' });
      // initialStoreState = { app: reportsReducer(undefined, { type: 'NONE' }) };
      configuredStore = configureStore([thunk]);
    });

    beforeEach(() => {
      // reset store state and clears actions
      store = configuredStore(initialStoreState);
    });

    function mockDispatchAction(state, action, propertyMatcher) {
      // slightly different implementation in this redux module test.
      const newStore = configuredStore(state);
      newStore.dispatch(action);
      const resolvedActions = newStore.getActions();
      let newState = state.layers;
      const actionPropertyMatcher = propertyMatcher ? { payload: propertyMatcher } : {};
      resolvedActions.forEach(resolvedAction => {
        newState = reportsReducer(newState, resolvedAction);
        expect(action).toMatchSnapshot(actionPropertyMatcher);
      });

      const statePropertyMatcher = propertyMatcher ? { list: propertyMatcher } : {};
      expect(newState).toMatchSnapshot(statePropertyMatcher);

      const returnState = {
        ...state,
        layers: {
          ...newState
        }
      };
      return returnState;
    }

    it('syncReports', () => {
      store.dispatch(syncReports()); // todo test sync/syncing
      expect(store.getActions()).toMatchSnapshot();
    });

    it('simple report actions full test', () => {
      let newState = {
        layers: reportsReducer(undefined, { type: 'NONE' }),
        // areas: {
        //   ...initialStoreState.areas,
        //   data: [mockArea]
        // }
      };
      const propertyMatcher = { mockCreateReportName: { date: expect.any(String) } };
      newState = mockDispatchAction(newState, createReport(mockCreateReport), propertyMatcher);
      newState = mockDispatchAction(newState, setOfflineMode(true));
      // mockDispatchAction(newState, showNotConnectedNotification());
    });

    // it('showNotConnectedNotification full test', () => {
    //   let newState = appReducer(undefined, { type: 'NONE' });
    //   newState = mockDispatchAction(newState, showNotConnectedNotification());
    //   newState = mockDispatchAction(newState, setOfflineMode(true));
    //   mockDispatchAction(newState, showNotConnectedNotification());
    // });
  });
});
