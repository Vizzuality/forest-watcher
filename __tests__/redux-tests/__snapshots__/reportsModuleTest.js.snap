// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Reports Module Initial reducer state 1`] = `
Object {
  "list": Object {},
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 1`] = `
Object {
  "payload": Object {
    "mockCreateReportName": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "type": "report/CREATE_REPORT",
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 2`] = `
Object {
  "list": Object {
    "mockCreateReportName": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 3`] = `
Object {
  "payload": Object {
    "mockCreateReportName2": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName2",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "type": "report/CREATE_REPORT",
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 4`] = `
Object {
  "list": Object {
    "mockCreateReportName": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName",
      "status": "draft",
      "userPosition": "1,2",
    },
    "mockCreateReportName2": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName2",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 5`] = `
Object {
  "payload": Object {
    "data": Object {
      "status": "complete",
    },
    "name": "mockCreateReportName2",
  },
  "type": "report/UPDATE_REPORT",
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 6`] = `
Object {
  "list": Object {
    "mockCreateReportName": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName",
      "status": "draft",
      "userPosition": "1,2",
    },
    "mockCreateReportName2": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName2",
      "status": "complete",
      "userPosition": "1,2",
    },
  },
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 7`] = `
Object {
  "payload": Object {
    "reportName": "mockCreateReportName",
  },
  "type": "report/DELETE_REPORT",
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions simple report actions full test 8`] = `
Object {
  "list": Object {
    "mockCreateReportName2": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName2",
      "status": "complete",
      "userPosition": "1,2",
    },
  },
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Redux Snapshot Thunk Actions syncReports 1`] = `
Array [
  Object {
    "meta": Object {
      "offline": Object {
        "commit": Object {
          "type": "report/GET_DEFAULT_TEMPLATE_COMMIT",
        },
        "effect": Object {
          "url": "<API_URL_MOCK>/reports/default",
        },
        "rollback": Object {
          "type": "report/GET_DEFAULT_TEMPLATE_ROLLBACK",
        },
      },
    },
    "type": "report/GET_DEFAULT_TEMPLATE_REQUEST",
  },
]
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test createReport 1`] = `
Object {
  "payload": Object {
    "mockCreateReportName": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "type": "report/CREATE_REPORT",
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test createReport 2`] = `
Object {
  "list": Object {
    "mockCreateReportName": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": Any<String>,
      "index": 0,
      "reportName": "mockCreateReportName",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test deleteReport 1`] = `
Object {
  "payload": Object {
    "reportName": "mockReportName",
  },
  "type": "report/DELETE_REPORT",
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test deleteReport 2`] = `
Object {
  "list": Object {},
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test getDefaultReport 1`] = `
Object {
  "meta": Object {
    "offline": Object {
      "commit": Object {
        "type": "report/GET_DEFAULT_TEMPLATE_COMMIT",
      },
      "effect": Object {
        "url": "<API_URL_MOCK>/reports/default",
      },
      "rollback": Object {
        "type": "report/GET_DEFAULT_TEMPLATE_ROLLBACK",
      },
    },
  },
  "type": "report/GET_DEFAULT_TEMPLATE_REQUEST",
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test getDefaultReport 2`] = `
Object {
  "list": Object {},
  "synced": false,
  "syncing": true,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test saveReport 1`] = `
Object {
  "payload": Object {
    "data": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": "mockDateString",
      "index": 1,
      "name": "reportNameMock",
      "status": "draft",
      "userPosition": "1,2",
    },
    "name": "reportNameMock",
  },
  "type": "report/UPDATE_REPORT",
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test saveReport 2`] = `
Object {
  "list": Object {
    "reportNameMock": Object {
      "answers": Array [],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": "mockDateString",
      "index": 1,
      "name": "reportNameMock",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test saveReport with answers 1`] = `
Object {
  "payload": Object {
    "data": Object {
      "answers": Array [
        Object {
          "child": Object {
            "questionName": "mockQuestionName2",
            "value": "Value2",
          },
          "questionName": "mockQuestionName1",
          "value": "Value1",
        },
      ],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": "mockDateString",
      "index": 1,
      "name": "reportNameMock",
      "status": "draft",
      "userPosition": "1,2",
    },
    "name": "reportNameMock",
  },
  "type": "report/UPDATE_REPORT",
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test saveReport with answers 2`] = `
Object {
  "list": Object {
    "reportNameMock": Object {
      "answers": Array [
        Object {
          "child": Object {
            "questionName": "mockQuestionName2",
            "value": "Value2",
          },
          "questionName": "mockQuestionName1",
          "value": "Value1",
        },
      ],
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "clickedPosition": "3,4",
      "date": "mockDateString",
      "index": 1,
      "name": "reportNameMock",
      "status": "draft",
      "userPosition": "1,2",
    },
  },
  "synced": false,
  "syncing": false,
  "templates": Object {},
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test setReportAnswer 1`] = `
Object {
  "payload": Object {
    "answer": Object {
      "child": Object {
        "questionName": "mockQuestionName2",
        "value": "Value2",
      },
      "questionName": "mockQuestionName1",
      "value": "Value1",
    },
    "reportName": "mockReportName",
    "updateOnly": true,
  },
  "type": "report/SET_REPORT_ANSWER",
}
`;

exports[`Redux Reports Module Simple Actions: snapshot and reducer test setReportAnswer 2`] = `Object {}`;
